File: LeetCodePython/ShortestPathInBinaryMatrix.py
"""
Given an n x n binary matrix grid, return the length of the shortest clear path in the matrix. If there is no clear path, return -1.

A clear path in a binary matrix is a path from the top-left cell (i.e., (0, 0)) to the bottom-right cell (i.e., (n - 1, n - 1)) such that:

All the visited cells of the path are 0.
All the adjacent cells of the path are 8-directionally connected (i.e., they are different and they share an edge or a corner).
The length of a clear path is the number of visited cells of this path.

Example 1:
Input: grid = [[0,1],[1,0]]
Output: 2

Example 2:
Input: grid = [[0,0,0],[1,1,0],[1,1,0]]
Output: 4

Example 3:
Input: grid = [[1,0,0],[1,1,0],[1,1,0]]
Output: -1

Constraints:

n == grid.length
n == grid[i].length
1 <= n <= 100
grid[i][j] is 0 or 1

hint:
1 Do a breadth first search to find the shortest path.
"""
import collections
from typing import List


class ShortestPathInBinaryMatrix:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0), (1, 1), (1, -1), (-1, 1), (-1, -1)]
        if grid[0][0] == 1 or grid[rows - 1][cols - 1] == 1:
            return -1
        step = 1
        visited = [[False] * cols for _ in range(rows)]
        q = collections.deque()
        q.append((0, 0))
        visited[0][0] = True
        while q:
            sz = len(q)
            for _ in range(sz):
                r, c = q.popleft()
                if r == rows - 1 and c == cols - 1:
                    return step
                for dr, dc in dirs:
                    nb_r, nb_c = r + dr, c + dc
                    if 0 <= nb_r < rows and 0 <= nb_c < cols:
                        if grid[nb_r][nb_c] == 0 and not visited[nb_r][nb_c]:
                            q.append((nb_r, nb_c));
                            visited[nb_r][nb_c] = True
            step += 1
        return -1

File: LeetCodePractice/src/bfs/ShortestPathInBinaryMatrix.java
package bfs;

import java.util.LinkedList;
import java.util.Queue;

/*
#1091
In an N by N square grid, each cell is either empty (0) or blocked (1).

A clear path from top-left to bottom-right has length k if and only if it is composed of cells C_1, C_2, ..., C_k such that:

Adjacent cells C_i and C_{i+1} are connected 8-directionally (ie., they are different and share an edge or corner)
C_1 is at location (0, 0) (ie. has value grid[0][0])
C_k is at location (N-1, N-1) (ie. has value grid[N-1][N-1])
If C_i is located at (r, c), then grid[r][c] is empty (ie. grid[r][c] == 0).
Return the length of the shortest such clear path from top-left to bottom-right.  If such a path does not exist, return -1.

Example 1:

Input: [[0,1],[1,0]]
Output: 2
Example 2:

Input: [[0,0,0],[1,1,0],[1,1,0]]
Output: 4

Note:
1 <= grid.length == grid[0].length <= 100
grid[r][c] is 0 or 1
*/

public class ShortestPathInBinaryMatrix {
    
    public int shortestPathBinaryMatrix(int[][] grid) {
        int[][] dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}};
        int rows = grid.length;
        int cols = grid[0].length;
        if (grid[0][0] == 1 || grid[rows - 1][cols - 1] == 1) {
            return -1;
        }
        int step = 1;
        boolean[][] visited = new boolean[rows][cols];
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0, 0});
        visited[0][0] = true;
        while (!queue.isEmpty()) {
            int sz = queue.size();
            for (int i = 0; i < sz; i++) {
                int[] cur = queue.poll();
                if (cur[0] == rows - 1 && cur[1] == cols - 1) {
                    return step;
                }
                for (int[] dir : dirs) {
                    int nb_x = cur[0] + dir[0];
                    int nb_y = cur[1] + dir[1];
                    if (nb_x >= 0 && nb_x < rows && nb_y >= 0 && nb_y < cols) {
                        if (grid[nb_x][nb_y] == 0 && !visited[nb_x][nb_y]) {
                            queue.offer(new int[]{nb_x, nb_y});
                            visited[nb_x][nb_y] = true;
                        }
                    }
                }
            }
            step++;
        }
        return -1;
    }
}
